subprojects/core/src/main/java/org/gradle/deployment/internal/DefaultDeploymentRegistry.java

```diff
@@ -17,61 +17,128 @@
+        this.pendingChangesManager = pendingChangesManager;
+        this.buildOperationExecutor = buildOperationExecutor;
+        this.objectFactory = objectFactory;
+        this.pendingChanges = new PendingChanges();
+        // TODO: Detangle pending changes handling and continuous build
+        if (startParameter.isContinuous()) {
+            pendingChanges.changesMade();
+        }
+        pendingChangesManager.addListener(this);
+    }
+
     @Override
-    public void register(String id, DeploymentHandle handle) {
+    public <T extends DeploymentHandle> T start(final String name, final Class<T> handleType, final Object... params) {
         lock.lock();
         try {
             failIfStopped();
-            if (!handles.containsKey(id)) {
-                handles.put(id, handle);
+            if (!handles.containsKey(name)) {
+                return buildOperationExecutor.call(new CallableBuildOperation<T>() {
+                    @Override
+                    public BuildOperationDescriptor.Builder description() {
+                        return BuildOperationDescriptor.displayName("Start deployment '" + name + "'");
+                    }
+
+                    @Override
+                    public T call(BuildOperationContext context) {
+                        T delegate = objectFactory.newInstance(handleType, params);
+                        DeploymentHandleWrapper handle = new DeploymentHandleWrapper(name, delegate);
+                        if (pendingChanges.hasRemainingChanges()) {
+                            handle.pendingChanges(true);
+                        }
+                        handles.put(name, handle);
+                        return delegate;
+                    }
+                });
             } else {
-                throw new IllegalStateException("A deployment with id '" + id + "' is already registered.");
+                throw new IllegalStateException("A deployment with id '" + name + "' is already registered.");
             }
         } finally {
             lock.unlock();
         }
     }
 
     @Override
-    public <T extends DeploymentHandle> T get(Class<T> handleType, String id) {
+    public <T extends DeploymentHandle> T get(String name, Class<T> handleType) {
         lock.lock();
         try {
             failIfStopped();
-            return Cast.cast(handleType, handles.get(id));
+            if (handles.containsKey(name)) {
+                return Cast.cast(handleType, handles.get(name).getDelegate());
+            } else {
+                return null;
+            }
         } finally {
             lock.unlock();
         }
     }
 
     @Override
-    public void onNewBuild(Gradle gradle) {
+    public void onPendingChanges() {
+        lock.lock();
+        try {
+            pendingChanges.changesMade();
+            for (DeploymentHandle handle : handles.values()) {
+                handle.pendingChanges(true);
```

Comment:
I think we can further simplify the `DeploymentHandle` API, based on the actions a _reloadable_ deployment would need to take. 
As a straw-man, why can't we limit it to:
- `deployment updated` : the handle should reload itself
- `deployment failed to build`: the handle should unload, and possibly provide a mechanism to show the failure (like Play).